presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestAddExchangesBelowPartialAggregationOverGroupIdRuleSet.java

```diff
@@ -0,0 +1,147 @@
+                .overrideStats(groupIdSourceId, PlanNodeStatsEstimate
+                        .builder()
+                        .setOutputRowCount(100_000_000)
+                        .addVariableStatistics(ImmutableMap.of(
+                                new VariableReferenceExpression(Optional.empty(), "groupingKey1", BIGINT), VariableStatsEstimate.builder().setDistinctValuesCount(groupingKey1NDV).build(),
+                                new VariableReferenceExpression(Optional.empty(), "groupingKey2", BIGINT), VariableStatsEstimate.builder().setDistinctValuesCount(groupingKey2NDV).build(),
+                                new VariableReferenceExpression(Optional.empty(), "groupingKey3", BIGINT), VariableStatsEstimate.builder().setDistinctValuesCount(groupingKey3NDV).build()))
+                        .build())
+                .on(p -> {
+                    VariableReferenceExpression groupingKey1 = p.variable("groupingKey1", BIGINT);
+                    VariableReferenceExpression groupingKey2 = p.variable("groupingKey2", BIGINT);
+                    VariableReferenceExpression groupingKey3 = p.variable("groupingKey3", BIGINT);
+                    VariableReferenceExpression groupId = p.variable("groupId", BIGINT);
+                    return p.exchange(
+                            exchangeBuilder -> exchangeBuilder
+                                    .scope(REMOTE_STREAMING)
+                                    .partitioningScheme(new PartitioningScheme(Partitioning.create(
+                                            FIXED_ARBITRARY_DISTRIBUTION,
+                                            ImmutableList.of()),
+                                            ImmutableList.copyOf(ImmutableList.of(groupingKey1, groupingKey2, groupingKey3, groupId))))
+                                    .addInputsSet(groupingKey1, groupingKey2, groupingKey3, groupId)
+                                    .addSource(
+                                            p.aggregation(builder -> builder
+                                                    .singleGroupingSet(groupingKey1, groupingKey2, groupingKey3, groupId)
+                                                    .step(PARTIAL)
+                                                    .source(p.groupId(
+                                                            ImmutableList.of(
+                                                                    ImmutableList.of(groupingKey1, groupingKey2),
+                                                                    ImmutableList.of(groupingKey1, groupingKey3)),
+                                                            ImmutableList.of(),
+                                                            groupId,
+                                                            p.values(new PlanNodeId(groupIdSourceId), groupingKey1, groupingKey2, groupingKey3))))));
+                });
+
+        if (partitionedBy.isEmpty()) {
```

Comment:
I don't like that the test itself has logic telling what to assert. Can we split these into separate tests and make the expected behavior clearer?